GenericWriter corrupts []byte DECIMAL values when decimal(...) resolves to FIXED_LEN_BYTE_ARRAY

## Summary

`parquet-go v0.29.0` appears to corrupt DECIMAL values when:

- the Go field type is `[]byte`
- the field has a `decimal(...)` tag
- the derived parquet schema resolves that field to `FIXED_LEN_BYTE_ARRAY`
- rows are written through the optimized `GenericWriter` / `GenericBuffer` path

In that case the written parquet values look like truncated slice-header bytes rather than the slice contents.

## Minimal repro

```go
package main

import (
	"bytes"
	"fmt"
	"io"

	"github.com/parquet-go/parquet-go"
)

type Row struct {
	Bet []byte `parquet:"bet_amount,decimal(8:29)"`
	Opt []byte `parquet:"total_bet_amount,optional,decimal(8:29)"`
}

func fixedTwo() []byte {
	b := make([]byte, 13)
	copy(b[9:], []byte{0x0b, 0xeb, 0xc2, 0x00}) // 200000000 in hex
	return b
}

func main() {
	rows := []Row{
		{Bet: fixedTwo(), Opt: fixedTwo()},
	}

	var out bytes.Buffer
	w := parquet.NewGenericWriter[Row](&out)

	if _, err := w.Write(rows); err != nil {
		panic(err)
	}
	if err := w.Close(); err != nil {
		panic(err)
	}

	f, err := parquet.OpenFile(bytes.NewReader(out.Bytes()), int64(out.Len()))
	if err != nil {
		panic(err)
	}

	fmt.Println(f.Schema())

	r := f.RowGroups()[0].Rows()
	defer r.Close()

	buf := make([]parquet.Row, 1)
	if _, err := r.ReadRows(buf); err != nil && err != io.EOF {
		panic(err)
	}

	for _, v := range buf[0] {
		fmt.Printf("column=%d bytes=%x\n", v.Column(), v.ByteArray())
	}
}
```

## Expected behavior

The schema is:

```text
required fixed_len_byte_array(13) bet_amount (DECIMAL(29,8));
optional fixed_len_byte_array(13) total_bet_amount (DECIMAL(29,8));
```

and both values should be the fixed-length big-endian representation of `2 * 10^8`, i.e.:

```text
0000000000000000000bebc200
```

which should read back as decimal value `2`.

## Actual behavior

The schema is indeed `FIXED_LEN_BYTE_ARRAY(13) DECIMAL(29,8)`, but the stored bytes are not the slice contents.

In my repro they look like this:

```text
40030100400100000d00000000
```

That decodes to garbage decimals instead of `2`.

When I change the field type from `[]byte` to `[13]byte`, the written values are correct.

## Why this looks like a library bug

From local inspection of `v0.29.0`:

1. `decimal(...)` on both `reflect.Array` and `reflect.Slice` resolves to `FixedLenByteArrayType(decimalFixedLenByteArraySize(precision))`.
2. `GenericWriter` uses the optimized `writeRowsFuncOf(...)` path when the schema derived from the Go type matches the configured schema.
3. For `[]byte`, that optimized path eventually reaches the fixed-len byte-array column buffer and bulk-copies raw row memory.
4. For a Go slice field, the raw row memory is the slice header, not the slice backing array.

So the writer appears to copy 13 bytes from the in-memory slice header rather than the 13 bytes referenced by the slice.

## Suspected code path

These locations are from `github.com/parquet-go/parquet-go v0.29.0`:

- `schema.go`: `decimal(...)` on `reflect.Slice` resolves to `FixedLenByteArrayType(...)`
- `writer.go`: optimized write path selected when the configured schema matches the type-derived schema
- `column_buffer_write.go`: `[]byte` is routed through the typed bulk-write path
- `column_buffer_fixed_len_byte_array.go`: fixed-len values are copied from the sparse row memory block

The observed bytes strongly resemble a truncated slice header: pointer / len / cap, not the actual payload.

## Workaround

I could avoid the corruption in two ways:

1. Use `[13]byte` instead of `[]byte`
2. Keep fixed arrays/pointers in the row type (without decimal tag), but provide an explicit schema with DECIMAL annotations so `GenericWriter` falls back to deconstructing rows instead of using the optimized direct-memory path

That workaround produces:

- correct `FIXED_LEN_BYTE_ARRAY(13) DECIMAL(29,8)` schema
- correct values
- nullable fields via `*[13]byte`

## Related limitation

There is also a related limitation with nullable fixed-len decimals:

```go
type Row struct {
	Value *[13]byte `parquet:"value,optional,decimal(8:29)"`
}
```

This panics during schema derivation because the `decimal(...)` tag handler rejects `reflect.Ptr`, even though plain `*[13]byte` without `decimal(...)` is supported.

That may deserve a separate fix / issue, but the main data-corruption issue above seems more urgent.

## Environment

- `github.com/parquet-go/parquet-go v0.29.0`
- Go `1.25.4`

Thank you for the quick response. I’ve tested your fix and can confirm that the written data is no longer corrupted.

However, I’ve encountered several other issues that are not directly related to data corruption, but rather to how decimals backed by byte slices are intended to work.

1. If byte slice is nil, i get the following panic
```
panic: cannot write byte slice of length 0 to FIXED_LEN_BYTE_ARRAY(13) column

goroutine 326 [running]:
github.com/parquet-go/parquet-go.writeRowsFuncOfByteSlice.func1({0x14004cec400, 0x1f, 0x0?}, {0x38?, 0x9e?, 0x1a?}, {{0x140001a9e38?, 0x1?, 0x230?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:451 +0x35c
github.com/parquet-go/parquet-go.writeRowsFuncOfOptional.func3({0x14004cec400, 0x1f, 0x20}, {0xa0?, 0xd8?, 0xce?}, {{0x140001a9e38?, 0x14004ee38c0?, 0x1c?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:368 +0x1d0
github.com/parquet-go/parquet-go.writeRowsFuncOfStruct.func2({0x14004cec400, 0x1f, 0x20}, {0x50?, 0x67?, 0xc8?}, {{0x140001a9d48?, 0x1c?, 0x14004c86740?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:631 +0x64
github.com/parquet-go/parquet-go.writeRowsFuncOfPointer.func1({0x14004cec400, 0x1f, 0x20}, {0xc0?, 0x38?, 0xee?}, {{0x14002e1c908?, 0x1400004e900?, 0x0?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:483 +0x9c
github.com/parquet-go/parquet-go.makeBufferFunc[...].func1({0x14002e1c908?, 0x59?, 0x0})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/buffer.go:87 +0x5c
github.com/parquet-go/parquet-go.(*GenericBuffer[...]).Write(...)
```

The whole idea of using a byte slice instead of a fixed-size array was to allow for a nullable decimal field. I would expect a null value to be written to the column in this case.

2. If the length of the byte slice does not match what is inferred from the decimal definition, the following panic occurs:
```
panic: cannot write byte slice of length 4 to FIXED_LEN_BYTE_ARRAY(13) column

goroutine 48 [running]:
github.com/parquet-go/parquet-go.writeRowsFuncOfByteSlice.func1({0x14004ea2c00, 0x1f, 0x20?}, {0xa0?, 0xd8?, 0x3b?}, {{0x14004a54320?, 0x140005fce80?, 0x1c?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:451 +0x35c
github.com/parquet-go/parquet-go.writeRowsFuncOfStruct.func2({0x14004ea2c00, 0x1f, 0x20}, {0xd0?, 0x71?, 0xa0?}, {{0x14004a54248?, 0x1c?, 0x14004a071c0?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:631 +0x64
github.com/parquet-go/parquet-go.writeRowsFuncOfPointer.func1({0x14004ea2c00, 0x1f, 0x20}, {0x80?, 0xce?, 0x5f?}, {{0x14004ea2800?, 0x14004899200?, 0x0?}})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/column_buffer_write.go:483 +0x9c
github.com/parquet-go/parquet-go.makeBufferFunc[...].func1({0x14004ea2800?, 0x3d?, 0x0})
	/Users/bond/.asdf/installs/golang/1.25.4/packages/pkg/mod/github.com/parquet-go/parquet-go@v0.29.1-0.20260410165623-d5de90093284/buffer.go:87 +0x5c
github.com/parquet-go/parquet-go.(*GenericBuffer[...]).Write(...)
```

I don’t have a deep understanding of the Parquet format, so I might be mistaken, but as far as I know, decimals can be stored not only as FIXED_LEN_BYTE_ARRAY, but also as BYTE_ARRAY.

It seems that a []byte type should map more naturally to BYTE_ARRAY. If FIXED_LEN_BYTE_ARRAY is required for decimals for any reason, I would expect shorter slices to be padded (e.g., with zeros) up to the required size.

At the same time, it would still make sense to panic if the provided slice exceeds the size defined by the decimal type in the schema.

--- a/column_buffer_write.go
+++ b/column_buffer_write.go
@@ -68,6 +68,15 @@ func writeRowsFuncOf(t reflect.Type, schema *Schema, path columnPath, tagReplace
 		return writeRowsFuncOfSmallInt(t, schema, path)
 	case reflect.Slice:
 		if t.Elem().Kind() == reflect.Uint8 {
+			// When the column physical type is FIXED_LEN_BYTE_ARRAY (e.g. the
+			// field carries a decimal(...) tag), the optimized direct-memory
+			// path would read the slice header bytes instead of the slice
+			// contents. Dispatch to a dedicated writer that indirects through
+			// the slice header. See issue #508.
+			column := schema.lazyLoadState().mapping.lookup(path)
+			if column.node.Type().Kind() == FixedLenByteArray {
+				return writeRowsFuncOfByteSlice(t, schema, path)
+			}
 			return writeRowsFuncOfRequired(t, schema, path)
 		} else {
 			return writeRowsFuncOfSlice(t, schema, path, tagReplacements)
@@ -400,6 +409,53 @@ func writeRowsFuncOfArray(t reflect.Type, schema *Schema, path columnPath) write
 		panic(fmt.Sprintf("cannot convert Go values of type "+typeNameOf(t)+" to FIXED_LEN_BYTE_ARRAY(%d)", columnLen))
 	}
 	return writeRowsFuncOfRequired(t, schema, path)
+}
+
+type byteSliceBuf struct{ values []byte }
+
+var byteSliceBufPool memory.Pool[byteSliceBuf]
+
+// writeRowsFuncOfByteSlice handles writing []byte Go fields into a
+// FIXED_LEN_BYTE_ARRAY column. The optimized path used for other []byte
+// columns treats each sparse.Array element as raw bytes, but for a slice
+// field that element is a slice header. This function reads each slice via
+// rows.StringArray() (string and []byte share the ptr/len header layout),
+// copies the referenced bytes into a contiguous scratch buffer, and forwards
+// the buffer as a flat sparse.Array so the column buffer's existing write
+// path sees real fixed-size elements.
+func writeRowsFuncOfByteSlice(t reflect.Type, schema *Schema, path columnPath) writeRowsFunc {
+	column := schema.lazyLoadState().mapping.lookup(path)
+	columnIndex := column.columnIndex
+	if columnIndex == math.MaxUint16 {
+		panic("parquet: column not found: " + path.String())
+	}
+	size := column.node.Type().Length()
+	return func(columns []ColumnBuffer, levels columnLevels, rows sparse.Array) {
+		n := rows.Len()
+		if n == 0 {
+			columns[columnIndex].writeValues(levels, rows)
+			return
+		}
+
+		buf := byteSliceBufPool.Get(
+			func() *byteSliceBuf { return new(byteSliceBuf) },
+			func(b *byteSliceBuf) { b.values = b.values[:0] },
+		)
+		buf.values = slices.Grow(buf.values, n*size)[:n*size]
+		defer byteSliceBufPool.Put(buf)
+
+		stringArray := rows.StringArray()
+		for i := range n {
+			s := stringArray.Index(i)
+			if len(s) != size {
+				panic(fmt.Sprintf("cannot write byte slice of length %d to FIXED_LEN_BYTE_ARRAY(%d) column", len(s), size))
+			}
+			copy(buf.values[i*size:], s)
+		}
+
+		flatArray := sparse.UnsafeArray(unsafe.Pointer(&buf.values[0]), n, uintptr(size))
+		columns[columnIndex].writeValues(levels, flatArray)
+	}
 }
 
 func writeRowsFuncOfPointer(t reflect.Type, schema *Schema, path columnPath, tagReplacements []StructTagOption) writeRowsFunc {

--- a/column_buffer_write.go
+++ b/column_buffer_write.go
@@ -442,15 +442,23 @@ func writeRowsFuncOfByteSlice(t reflect.Type, schema *Schema, path columnPath) w
 			func(b *byteSliceBuf) { b.values = b.values[:0] },
 		)
 		buf.values = slices.Grow(buf.values, n*size)[:n*size]
+		// Zero-fill so empty/nil slice slots emit placeholder bytes; the
+		// pooled buffer may otherwise carry stale contents from a previous
+		// caller. The optional wrapper upstream marks those rows null via
+		// the definition level, matching fixedLenByteArrayColumnBuffer.writeNull.
+		clear(buf.values)
 		defer byteSliceBufPool.Put(buf)
 
 		stringArray := rows.StringArray()
 		for i := range n {
 			s := stringArray.Index(i)
-			if len(s) != size {
+			switch len(s) {
+			case 0:
+			case size:
+				copy(buf.values[i*size:], s)
+			default:
 				panic(fmt.Sprintf("cannot write byte slice of length %d to FIXED_LEN_BYTE_ARRAY(%d) column", len(s), size))
 			}
-			copy(buf.values[i*size:], s)
 		}
 
 		flatArray := sparse.UnsafeArray(unsafe.Pointer(&buf.values[0]), n, uintptr(size))

--- a/schema.go
+++ b/schema.go
@@ -1100,19 +1100,30 @@ func makeNodeOf(path []string, t reflect.Type, name string, tags parquetTags, ta
 				if err != nil {
 					throwInvalidTag(t, name, option+args)
 				}
-				var baseType Type
-				switch t.Kind() {
-				case reflect.Int32:
-					baseType = Int32Type
-				case reflect.Int64:
-					baseType = Int64Type
-				case reflect.Array, reflect.Slice:
-					baseType = FixedLenByteArrayType(decimalFixedLenByteArraySize(precision))
-				default:
-					throwInvalidTag(t, name, option)
+				decimalBaseType := func(et reflect.Type) Type {
+					switch et.Kind() {
+					case reflect.Int32:
+						return Int32Type
+					case reflect.Int64:
+						return Int64Type
+					case reflect.Array, reflect.Slice:
+						return FixedLenByteArrayType(decimalFixedLenByteArraySize(precision))
+					}
+					return nil
 				}
-
-				setNode(Decimal(scale, precision, baseType))
+				if t.Kind() == reflect.Ptr {
+					baseType := decimalBaseType(t.Elem())
+					if baseType == nil {
+						throwInvalidTag(t, name, option)
+					}
+					setNode(Optional(Decimal(scale, precision, baseType)))
+				} else {
+					baseType := decimalBaseType(t)
+					if baseType == nil {
+						throwInvalidTag(t, name, option)
+					}
+					setNode(Decimal(scale, precision, baseType))
+				}
 
 			case "string":
 				switch {

--- a/writer_test.go
+++ b/writer_test.go
@@ -3464,6 +3464,124 @@ func TestIssue449DecimalReadWrite(t *testing.T) {
 	}
 }
 
+// TestIssue508ByteSliceDecimalReadWrite reproduces issue #508: when a Go
+// struct field is declared as []byte with a decimal(...) tag that resolves to
+// FIXED_LEN_BYTE_ARRAY, the optimized GenericWriter path used to copy the
+// slice header bytes instead of the slice contents. This test round-trips
+// multiple rows with distinct values through GenericWriter and asserts both
+// the required and optional columns contain the exact input bytes. Multiple
+// rows are important: a single-row test would miss off-by-one errors in the
+// byte-buffer indexing of the fix.
+func TestIssue508ByteSliceDecimalReadWrite(t *testing.T) {
+	type Row struct {
+		Bet []byte `parquet:"bet_amount,decimal(8:29)"`
+		Opt []byte `parquet:"total_bet_amount,optional,decimal(8:29)"`
+	}
+
+	// Encode `value` big-endian into a fresh 13-byte fixed-length buffer.
+	encode := func(value uint32) []byte {
+		b := make([]byte, 13)
+		b[9] = byte(value >> 24)
+		b[10] = byte(value >> 16)
+		b[11] = byte(value >> 8)
+		b[12] = byte(value)
+		return b
+	}
+
+	rows := []Row{
+		{Bet: encode(200000000), Opt: encode(200000000)},
+		{Bet: encode(1), Opt: encode(2)},
+		{Bet: encode(0xdeadbeef), Opt: encode(0xcafef00d)},
+	}
+
+	var buf bytes.Buffer
+	w := parquet.NewGenericWriter[Row](&buf)
+	if _, err := w.Write(rows); err != nil {
+		t.Fatalf("unable to write: %s", err)
+	}
+	if err := w.Close(); err != nil {
+		t.Fatalf("unable to close writer: %s", err)
+	}
+
+	f, err := parquet.OpenFile(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
+	if err != nil {
+		t.Fatalf("unable to open file: %s", err)
+	}
+
+	for _, col := range []string{"bet_amount", "total_bet_amount"} {
+		leaf, ok := f.Schema().Lookup(col)
+		if !ok {
+			t.Fatalf("column %q not found in schema", col)
+		}
+		if got, want := leaf.Node.Type().Kind(), parquet.FixedLenByteArray; got != want {
+			t.Fatalf("column %q: expected physical type %v, got %v", col, want, got)
+		}
+		if got, want := leaf.Node.Type().Length(), 13; got != want {
+			t.Fatalf("column %q: expected length %d, got %d", col, want, got)
+		}
+	}
+
+	// Use the raw row reader so we can assert on the exact bytes stored in
+	// the parquet file, bypassing the decimal->*big.Float reconstruct path
+	// which does not support []byte destinations.
+	reader := f.RowGroups()[0].Rows()
+	defer reader.Close()
+
+	rowBuf := make([]parquet.Row, len(rows))
+	n, err := reader.ReadRows(rowBuf)
+	if err != nil && err != io.EOF {
+		t.Fatalf("unable to read rows: %s", err)
+	}
+	if n != len(rows) {
+		t.Fatalf("expected %d rows, got %d", len(rows), n)
+	}
+
+	for i, got := range rowBuf {
+		if len(got) != 2 {
+			t.Fatalf("row %d: expected 2 column values, got %d", i, len(got))
+		}
+		if !bytes.Equal(got[0].ByteArray(), rows[i].Bet) {
+			t.Errorf("row %d bet_amount: got %x, want %x", i, got[0].ByteArray(), rows[i].Bet)
+		}
+		if !bytes.Equal(got[1].ByteArray(), rows[i].Opt) {
+			t.Errorf("row %d total_bet_amount: got %x, want %x", i, got[1].ByteArray(), rows[i].Opt)
+		}
+	}
+}
+
+// TestIssue508ByteSliceDecimalWrongLength asserts that writing a []byte value
+// whose length does not match the FIXED_LEN_BYTE_ARRAY size fails with a
+// clear error instead of silently truncating or corrupting the output. Before
+// the fix, wrong-length input was copied as slice-header bytes and produced
+// plausible-looking but incorrect column data.
+func TestIssue508ByteSliceDecimalWrongLength(t *testing.T) {
+	type Row struct {
+		Bet []byte `parquet:"bet_amount,decimal(8:29)"`
+	}
+
+	defer func() {
+		r := recover()
+		if r == nil {
+			t.Fatal("expected panic on wrong-length []byte value, got none")
+		}
+		msg, ok := r.(string)
+		if !ok {
+			if e, isErr := r.(error); isErr {
+				msg = e.Error()
+			}
+		}
+		if !strings.Contains(msg, "FIXED_LEN_BYTE_ARRAY(13)") {
+			t.Fatalf("unexpected panic message: %v", r)
+		}
+	}()
+
+	var buf bytes.Buffer
+	w := parquet.NewGenericWriter[Row](&buf)
+	// 5 bytes instead of the required 13 — must fail loudly.
+	_, _ = w.Write([]Row{{Bet: []byte{1, 2, 3, 4, 5}}})
+	_ = w.Close()
+}
+
 func TestWriteOptionalJSONRawMessage(t *testing.T) {
 	type Row struct {
 		Buf json.RawMessage `parquet:"buf,json,optional"`

--- a/writer_test.go
+++ b/writer_test.go
@@ -3582,6 +3582,161 @@ func TestIssue508ByteSliceDecimalWrongLength(t *testing.T) {
 	_ = w.Close()
 }
 
+// TestIssue508OptionalNilByteSliceWritesNull covers the second issue surfaced
+// in the #508 thread: an optional []byte field that resolves to
+// FIXED_LEN_BYTE_ARRAY used to panic with "cannot write byte slice of length
+// 0" when the slice was nil. A nil value on an optional field should write
+// NULL instead.
+func TestIssue508OptionalNilByteSliceWritesNull(t *testing.T) {
+	type Row struct {
+		Opt []byte `parquet:"value,optional,decimal(8:29)"`
+	}
+
+	encode := func(value uint32) []byte {
+		b := make([]byte, 13)
+		b[9] = byte(value >> 24)
+		b[10] = byte(value >> 16)
+		b[11] = byte(value >> 8)
+		b[12] = byte(value)
+		return b
+	}
+
+	rows := []Row{
+		{Opt: encode(1)},
+		{Opt: nil},
+		{Opt: encode(2)},
+		{Opt: nil},
+	}
+
+	var buf bytes.Buffer
+	w := parquet.NewGenericWriter[Row](&buf)
+	if _, err := w.Write(rows); err != nil {
+		t.Fatalf("unable to write: %s", err)
+	}
+	if err := w.Close(); err != nil {
+		t.Fatalf("unable to close writer: %s", err)
+	}
+
+	f, err := parquet.OpenFile(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
+	if err != nil {
+		t.Fatalf("unable to open file: %s", err)
+	}
+
+	reader := f.RowGroups()[0].Rows()
+	defer reader.Close()
+
+	rowBuf := make([]parquet.Row, len(rows))
+	n, err := reader.ReadRows(rowBuf)
+	if err != nil && err != io.EOF {
+		t.Fatalf("unable to read rows: %s", err)
+	}
+	if n != len(rows) {
+		t.Fatalf("expected %d rows, got %d", len(rows), n)
+	}
+
+	for i, got := range rowBuf {
+		if len(got) != 1 {
+			t.Fatalf("row %d: expected 1 column value, got %d", i, len(got))
+		}
+		if rows[i].Opt == nil {
+			if !got[0].IsNull() {
+				t.Errorf("row %d: expected null, got bytes %x", i, got[0].ByteArray())
+			}
+		} else {
+			if got[0].IsNull() {
+				t.Errorf("row %d: expected non-null, got null", i)
+			}
+			if !bytes.Equal(got[0].ByteArray(), rows[i].Opt) {
+				t.Errorf("row %d: got %x, want %x", i, got[0].ByteArray(), rows[i].Opt)
+			}
+		}
+	}
+}
+
+// TestIssue508PointerArrayDecimalSupported covers the third issue from the
+// #508 thread: *[N]byte with a decimal(...) tag used to panic during schema
+// derivation because the decimal tag handler did not unwrap reflect.Ptr. A
+// pointer to a fixed-length byte array is the natural way to express a
+// nullable fixed-length decimal column.
+func TestIssue508PointerArrayDecimalSupported(t *testing.T) {
+	type Row struct {
+		Value *[13]byte `parquet:"value,decimal(8:29)"`
+	}
+
+	encode := func(value uint32) *[13]byte {
+		var b [13]byte
+		b[9] = byte(value >> 24)
+		b[10] = byte(value >> 16)
+		b[11] = byte(value >> 8)
+		b[12] = byte(value)
+		return &b
+	}
+
+	rows := []Row{
+		{Value: encode(1)},
+		{Value: nil},
+		{Value: encode(0xdeadbeef)},
+	}
+
+	var buf bytes.Buffer
+	w := parquet.NewGenericWriter[Row](&buf)
+	if _, err := w.Write(rows); err != nil {
+		t.Fatalf("unable to write: %s", err)
+	}
+	if err := w.Close(); err != nil {
+		t.Fatalf("unable to close writer: %s", err)
+	}
+
+	f, err := parquet.OpenFile(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
+	if err != nil {
+		t.Fatalf("unable to open file: %s", err)
+	}
+
+	leaf, ok := f.Schema().Lookup("value")
+	if !ok {
+		t.Fatal("column \"value\" not found in schema")
+	}
+	if got, want := leaf.Node.Type().Kind(), parquet.FixedLenByteArray; got != want {
+		t.Fatalf("expected physical type %v, got %v", want, got)
+	}
+	if got, want := leaf.Node.Type().Length(), 13; got != want {
+		t.Fatalf("expected length %d, got %d", want, got)
+	}
+	if !leaf.Node.Optional() {
+		t.Fatal("expected pointer field to derive an optional schema node")
+	}
+
+	reader := f.RowGroups()[0].Rows()
+	defer reader.Close()
+
+	rowBuf := make([]parquet.Row, len(rows))
+	n, err := reader.ReadRows(rowBuf)
+	if err != nil && err != io.EOF {
+		t.Fatalf("unable to read rows: %s", err)
+	}
+	if n != len(rows) {
+		t.Fatalf("expected %d rows, got %d", len(rows), n)
+	}
+
+	for i, got := range rowBuf {
+		if len(got) != 1 {
+			t.Fatalf("row %d: expected 1 column value, got %d", i, len(got))
+		}
+		if rows[i].Value == nil {
+			if !got[0].IsNull() {
+				t.Errorf("row %d: expected null, got bytes %x", i, got[0].ByteArray())
+			}
+		} else {
+			if got[0].IsNull() {
+				t.Errorf("row %d: expected non-null, got null", i)
+			}
+			if !bytes.Equal(got[0].ByteArray(), rows[i].Value[:]) {
+				t.Errorf("row %d: got %x, want %x", i, got[0].ByteArray(), rows[i].Value[:])
+			}
+		}
+	}
+}
+
 func TestWriteOptionalJSONRawMessage(t *testing.T) {
 	type Row struct {
 		Buf json.RawMessage `parquet:"buf,json,optional"`

